src/engine/readaheadmanager.cpp

```diff
@@ -23,88 +25,113 @@ ReadAheadManager::ReadAheadManager(CachingReader* pReader,
 
-    if (samples_read != samples_needed) {
-        qDebug() << "didn't get what we wanted" << samples_read << samples_needed;
+    if (samples_read != samples_from_reader) {
+        qDebug() << "didn't get what we wanted" << samples_read << samples_from_reader;
     }
 
     // Increment or decrement current read-ahead position
+    // Mixing int and double here is desired, because the fractional frame should
+    // be resist
     if (in_reverse) {
-        addReadLogEntry(m_iCurrentPosition, m_iCurrentPosition - samples_read);
-        m_iCurrentPosition -= samples_read;
+        addReadLogEntry(m_currentPosition, m_currentPosition - samples_read);
+        m_currentPosition -= samples_read;
     } else {
-        addReadLogEntry(m_iCurrentPosition, m_iCurrentPosition + samples_read);
-        m_iCurrentPosition += samples_read;
+        addReadLogEntry(m_currentPosition, m_currentPosition + samples_read);
+        m_currentPosition += samples_read;
     }
 
     // Activate on this trigger if necessary
     if (loop_active) {
         // LoopingControl makes the decision about whether we should loop or
         // not.
-        const double loop_target = m_pLoopingControl->
-                process(dRate, m_iCurrentPosition, 0, 0);
+        const double loop_target = m_pLoopingControl->process(
+                dRate, m_currentPosition, 0, 0);
 
         if (loop_target != kNoTrigger) {
-            m_iCurrentPosition = loop_target;
+            m_currentPosition = loop_target;
+            if (preloop_samples > 0) {
+                // we are up to one frame ahead of the loop trigger
+                double overshoot = preloop_samples - samplesToLoopTrigger;
+                // start the loop later accordingly to be sure the loop length is as desired
+                // e.g. exactly one bar.
+                m_currentPosition += overshoot;
+
+                // Example in frames;
+                // loop start 1.1 loop end 3.3 loop length 2.2
+                // m_currentPosition samplesToLoopTrigger preloop_samples
+                // 2.0               1.3                  2
+                // 1.8               1.5                  2
+                // 1.6               1.7                  2
+                // 1.4               1.9                  2
+                // 1.2               2.1                  3
+                // Average preloop_samples = 2.2
+            }
 
-            int loop_read_position = m_iCurrentPosition +
-                    (in_reverse ? preloop_samples : -preloop_samples);
+            // start reading before the loop start point, to crossfade these samples
+            // with the samples we need to the loop end
+            int loop_read_position = SampleUtil::roundPlayPosToFrameStart(
+                    m_currentPosition + (in_reverse ? preloop_samples : -preloop_samples), kNumChannels);
 
             int looping_samples_read = m_pReader->read(
```

Comment:
Here you still use the old way of passing the first sample to the reader. Maybe it's safer to revert what I told you and always specify the first sample (= anchor) depending on the direction. This feels more "correct" to me now that I look at the resulting code. Doing lots of conditional computations depending on reverse or not is more a deficiency of the ReadAheadManager and a missing abstraction for handling directional audio data (see above, we need to encapsulate all this into a class sometime)